### PR TITLE
Add utilities for common file operations

### PR DESCRIPTION
**Pull request by nivlac12**
This PR adds generic functions for file operations in `file_utils.py`.

### CODE REVIEW

File: python/vsi/utils/file_utils.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+import re
+import shutil
+import stat
+import tempfile
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+def rmtree(path, ignore_errors=False, onerror=None, rmdir=True):
```
Comment (drewgilliam) [inline python/vsi/utils/file_utils.py:15]: Please add docstring

File: python/vsi/utils/file_utils.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        except OSError:
+          onerror(os.path.islink, path, sys.exc_info())
+    finally:
+      os.close(fd)
+  else:
+    return _rmtree_unsafe(path, onerror, rmdir)
+
+
+# version vulnerable to race conditions
+def _rmtree_unsafe(path, onerror, rmdir=True):
```
Comment (drewgilliam) [inline python/vsi/utils/file_utils.py:60]: Please add docstring

File: python/vsi/utils/file_utils.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
+""" Utility Functions for common file operations """
```
Comment (drewgilliam) [inline python/vsi/utils/file_utils.py:1]: Are there unit tests that go along with these functions?

Comment (nivlac12) [reply python/vsi/utils/file_utils.py:1]: Added `test_file_utils.py`, and confirmed that it passes locally.

File: python/vsi/test/test_file_utils.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      foo_dict = {1: 'test'}
+      with open(foo_json_file, 'w') as fp:
+        json.dump(foo_dict, fp)
+
+      # test for image neighbors of json file
+      src_file = foo_json_file
+      image_extensions = ('tif', 'tiff')
+      support_files = file_utils.get_neighbor_files_with_extension(src_file, image_extensions)
+      expected_file = [foo_image_file]
+      self.assertEqual(support_files, expected_file)
```
Comment (drewgilliam) [inline python/vsi/test/test_file_utils.py:124]: Should each of these be within a `subTest` to provide better reporting of a failure?

https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

File: python/vsi/utils/file_utils.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+      # defined file - call with original kwargs
+      else:
+        return func(*args, **kwargs)
+
+    return wrapper
+  return inner
+
+
+def dir_or_temp(kwarg_key):
```
Comment (drewgilliam) [inline python/vsi/utils/file_utils.py:233]: I don't see any unit tests for this function, can unit tests be added?

File: python/vsi/utils/file_utils.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+  defines a directory with a valid temporary directory.
+
+  The temporary directory will be created on entering the decorated function,
+  and the directory & its contents will be deleted on exiting the decorated
+  function.
+  '''
+  return _path_or_temp(kwarg_key)
+
+
+def file_or_temp(kwarg_key, filename):
```
Comment (drewgilliam) [inline python/vsi/utils/file_utils.py:245]: I don't see any unit tests for this function, can unit tests be added?

File: python/vsi/test/test_file_utils.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      # create some foo files with the same name, but different extensions
+      # a text file
+      foo_text_file = os.path.join(foo_dir, 'temp1.txt')
+      with open(foo_text_file, 'w') as foo:
+        foo.write('test')
+
+      # an image file
+      foo_image_file = os.path.join(foo_dir, 'temp1.tif')
+      foo_data = np.random.randint(0, 255, (256, 256, 3), 'uint8')
+      imwrite(foo_image_file, foo_data, photometric='rgb')
```
Comment (andyneff) [inline python/vsi/test/test_file_utils.py:111]: Does this need to be a real tiff file for this test? Can't it just be an empty file?

File: python/vsi/test/test_file_utils.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+foo_files = ['foo_file0.txt', 'foo_file1.txt', 'foo_file2.txt', 'foo_f0.txt','foo_f1.txt', 'f2.txt']
+foo_pattern_files = ['foo_08_04_2023.txt', 'foo_08_04_23.txt']
+
+class TestFileUtils(TestCase):
+
+  def test_rmtree(self):
+   # test remove tree with the ability to keep the top dir
+
+   # write temp subdirectories to a temp directory
+   with tempfile.TemporaryDirectory() as foo_dir:
```
[the remark this answers is not in the document]
Comment (nivlac12) [reply python/vsi/test/test_file_utils.py]: Removed `tempfile.TemporaryDirectory()` calls, used `self.temp_dir.name` instead

File: python/vsi/utils/file_utils.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+          onerror(os.path.islink, path, sys.exc_info())
+    finally:
+      os.close(fd)
+  else:
+    return _rmtree_unsafe(path, onerror, rmdir)
+
+
+# version vulnerable to race conditions
+def _rmtree_unsafe(path, onerror, rmdir=True):
+  """" Version of rmtree vulnerable to race conditions.
```
[the remark this answers is not in the document]
Comment (scott-vsi) [reply python/vsi/utils/file_utils.py:74]: that would be worth documenting. I assume you mean both rmtree and rmtree_unsafe?

(I see that this version is used by `rmtree` if shutil._use_fd_functions is False, but I don't really get what would cause shutil._use_fd_functions to be False.)

[the remark this answers is not in the document]
Comment (nivlac12) [reply python/vsi/utils/file_utils.py:74]: Since `_rmtree_unsafe` is part of how `shutil` performs `rmtree` (https://github.com/python/cpython/blob/main/Lib/shutil.py#L662), I think we should leave it in. @scott-vsi, if you agree you can resolve this thread.

File: python/vsi/utils/file_utils.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+          try:
+            os.rmdir(path)
+          except OSError:
+            onerror(os.rmdir, path, sys.exc_info())
+      else:
+        try:
+          # symlinks to directories are forbidden, see bug #1669
+          raise OSError("Cannot call rmtree on a symbolic link")
+        except OSError:
+          onerror(os.path.islink, path, sys.exc_info())
```
[the remark this answers is not in the document]
Comment (scott-vsi) [reply python/vsi/utils/file_utils.py:65]: this is done in cpython:shutil, from which this was copied...

File: python/vsi/utils/file_utils.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+      raise
+  if shutil._use_fd_functions:
+    # While the unsafe rmtree works fine on bytes, the fd based does not.
+    if isinstance(path, bytes):
+      path = os.fsdecode(path)
+    # Note: To guard against symlink races, we use the standard
+    # lstat()/open()/fstat() trick.
+    try:
+      orig_st = os.lstat(path)
+    except Exception:
```
[the remark this answers is not in the document]
Comment (scott-vsi) [reply python/vsi/utils/file_utils.py:44]: again, this is done in cpython (amazingly) up till [v3.12.0](https://github.com/python/cpython/blob/v3.12.0/Lib/shutil.py#L704)

File: python/vsi/utils/file_utils.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+      except OSError:
+        onerror(os.unlink, fullname, sys.exc_info())
+  if rmdir:
+    try:
+      os.rmdir(path)
+    except OSError:
+      onerror(os.rmdir, path, sys.exc_info())
+
+
+def get_files_with_extension_from_dir(directory, extensions):
```
[the remark this answers is not in the document]
Comment (nivlac12) [reply python/vsi/utils/file_utils.py]: Changed name, added the ability to find recursively or base-level only, and added test case.

File: python/vsi/utils/file_utils.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+import re
+import shutil
+import stat
+import tempfile
+
+import logging
+logger = logging.getLogger(__name__)
+
+# This function was copied from cpython,
+# and changed to allow the root directory to be kept.
```
Comment (scott-vsi) [inline python/vsi/utils/file_utils.py:15]:
```suggestion
# and changed to allow the base directory to be kept.
```

File: python/vsi/utils/file_utils.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+  ----------
+  path : str
+      Path to the directory with contents to remove
+  ignore_errors : bool, optional
+      If True, exceptions will not be raised when an error is encountered. Default: False
+  onerror : Callable, optional
+      Function to be called when an error is encountered. Only used if ignore_errors is False.
+      Default: None
+  keep_base_dir : bool, optional
+      If True, the directory will remain while its children are deleted. Default: False
```
Comment (scott-vsi) [inline python/vsi/utils/file_utils.py:32]:
```suggestion
      If True, the base directory will remain while its children are deleted. Default: False
```

**Review by scott-vsi — APPROVED** [review]

File: python/vsi/utils/file_utils.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+  Parameters
+  ----------
+  path : str
+      Path to the directory with contents to remove
+  ignore_errors : bool, optional
+      True if exceptions should not be raised when an error is encountered. Default: False
+  onerror : Callable, optional
+      Function to be called when an error is encountered. Only used if ignore_errors is false.
+      Default: None
+  rmdir : bool, optional
```
[the remark this answers is not in the document]
Comment (nivlac12) [reply python/vsi/utils/file_utils.py]: Changed to `keep_base_dir`.

File: python/vsi/utils/file_utils.py [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+  '''
+
+  # Convert the extensions to lower case
+  lower_case_extensions = [ext.lower() for ext in extensions]
+
+  # List of files to return
+  files = []
+
+  # For every file in directory
+  for f in glob.glob(os.path.join(directory, "**", "*.*"), recursive=True):
```
[the remark this answers is not in the document]
Comment (nivlac12) [reply python/vsi/utils/file_utils.py]: `glob.glob` seems quite fast, so leaving as is.
